subsys/bluetooth/host/rfcomm.c

```diff
@@ -535,11 +553,103 @@ static void rfcomm_check_fc(struct bt_rfcomm_dlc *dlc)
 	k_sem_give(&dlc->tx_credits);
 }
 
+static struct bt_rfcomm_tx *bt_rfcomm_tx_alloc(void)
+{
+	/* The TX context always get freed in the system workqueue,
+	 * so if we're in the same workqueue but there are no immediate
+	 * contexts available, there's no chance we'll get one by waiting.
+	 */
+	if (k_current_get() == &k_sys_work_q.thread) {
+		return k_fifo_get(&rfcomm_tx_free, K_NO_WAIT);
+	}
+
+	if (IS_ENABLED(CONFIG_BT_RFCOMM_LOG_LEVEL_DBG)) {
+		struct bt_rfcomm_tx *tx = k_fifo_get(&rfcomm_tx_free, K_NO_WAIT);
+
+		if (tx) {
+			return tx;
+		}
+
+		LOG_WRN("Unable to get an immediate free bt_rfcomm_tx");
+	}
+
+	return k_fifo_get(&rfcomm_tx_free, K_FOREVER);
+}
+
+static void bt_rfcomm_tx_free(struct bt_rfcomm_tx *tx)
+{
+	LOG_DBG("Free tx buffer %p", tx);
+
+	tx->buf = NULL;
+	tx->dlc = NULL;
+	k_fifo_put(&rfcomm_tx_free, tx);
+}
+
+static void bt_rfcomm_tx_destroy(struct net_buf *buf)
+{
+	struct bt_rfcomm_tx *tx;
+	struct bt_rfcomm_dlc *dlc;
+	struct net_buf *tx_buf;
+
+	LOG_DBG("buf %p", buf);
+
+	if ((buf == NULL) || (buf->len == 0)) {
+		return;
+	}
+
+	tx = *rfcomm_tx_data(buf);
+	tx_buf = tx->buf;
+	dlc = tx->dlc;
+
+	bt_rfcomm_tx_free(tx);
+
+	if (tx_buf != buf) {
+		LOG_ERR("Tx buf %p and buf %p are inconsistent", tx_buf, buf);
+	}
+
+	if (dlc && dlc->ops && dlc->ops->sent) {
+		dlc->ops->sent(dlc, tx_buf, -ESHUTDOWN);
+	}
+
+	if (tx_buf != NULL) {
+		net_buf_unref(tx_buf);
+	}
+}
+
+static void rfcomm_sent(struct bt_conn *conn, void *user_data, int err)
+{
+	struct bt_rfcomm_tx *tx;
+	struct bt_rfcomm_dlc *dlc;
+	struct net_buf *tx_buf;
+
+	LOG_DBG("conn %p", conn);
+
+	if (user_data == NULL)	{
+		return;
+	}
+
+	tx = (struct bt_rfcomm_tx *)user_data;
```

Comment:
Minor: unnecessary typecast (void pointers are automatically compatible with any other pointer, except when going non-const to const of vice-versa)

Reply:
Got it.